cpp/oneapi/dal/algo/shortest_paths/backend/cpu/service.hpp

```diff
@@ -0,0 +1,348 @@
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#pragma once
+
+#include <atomic>
+
+#include "oneapi/dal/algo/shortest_paths/common.hpp"
+#include "oneapi/dal/backend/common.hpp"
+#include "oneapi/dal/backend/memory.hpp"
+#include "oneapi/dal/backend/interop/common.hpp"
+#include "oneapi/dal/detail/threading.hpp"
+
+namespace oneapi::dal::preview::shortest_paths::backend {
+using namespace oneapi::dal::preview::detail;
+using namespace oneapi::dal::preview::backend;
+
+namespace mode {
+struct distances;
+struct distances_predecessors;
+} // namespace mode
+
+template <typename EV, typename VT>
+struct dist_pred {
+    using distance_type = EV;
+    using vertex_type = VT;
+    dist_pred(const EV& dist_, const VT& pred_) : dist(dist_), pred(pred_) {}
+    EV dist = 0;
+    VT pred = 0;
+};
+
+template <class T1, class T2>
+bool operator==(const dist_pred<T1, T2>& lhs, const dist_pred<T1, T2>& rhs) {
+    return lhs.dist == rhs.dist && lhs.pred == rhs.pred;
+}
+
+template <class T1, class T2>
+bool operator!=(const dist_pred<T1, T2>& lhs, const dist_pred<T1, T2>& rhs) {
+    return !(lhs == rhs);
+}
+
+template <typename T1, typename T2, typename T1_, typename T2_>
+inline constexpr bool is_class_params_the_same_v =
+    std::conjunction_v<std::is_same<T1, T2>, std::is_same<T1, T1_>, std::is_same<T2, T2_>>;
+
+template <typename T1, typename T2, typename T1_, typename T2_>
+inline constexpr bool is_not_class_params_the_same_v =
+    !std::is_same_v<T1, T2> && std::is_same_v<T1, T1_> && std::is_same_v<T2, T2_>;
+
+template <typename Cpu, typename Mode, typename Vertex, typename Value, typename AtomicT>
+class data_to_relax_base {
+public:
+    using relaxing_data_type = Value;
+    using atomic_type = std::atomic<AtomicT>;
+    using atomic_value_type = AtomicT;
+    using atomic_value_allocator_type = inner_alloc<atomic_type>;
+    using vertex_type = Vertex;
+    using mode_type = Mode;
+
+    template <typename M = Mode, std::enable_if_t<is_same_v<M, mode::distances>, bool> = true>
+    data_to_relax_base(std::int64_t vertex_count,
+                       Vertex source,
+                       relaxing_data_type max_dist,
+                       byte_alloc_iface* alloc_ptr)
+            : vertex_count(vertex_count),
+              atomic_value_allocator(alloc_ptr) {
+        relaxing_data = allocate(atomic_value_allocator, vertex_count);
+        relaxing_data = new (relaxing_data) atomic_type[vertex_count]();
+        dal::detail::threader_for(vertex_count, vertex_count, [&](std::int64_t i) {
+            store_by_index(i, max_dist);
```

Comment:
won't you construct the value in parallel? and avoid code in 82 row
`new (relaxing_data[i]) atomic_type(max_dist);`

Reply:
Thanks. Need to update as such initialization atomic<T> obj(t) is not atomic and we will reduce the amount of atomics ops